### PR TITLE
feat: Add only-users functionality

### PR DESCRIPTION
**Pull request by jdcargile**
@singingwolfboy, I have added a feature for users of your GitHub Action create-linked-clubhouse-story to use a new input attribute called `included-users` where they can pass a comma-separated list of GitHub users. For example, if the author of the current PR is in the `included-users` list, it will create a linked Clubhouse story. This is handy if people using this action only want automated Clubhouse stories created for certain users. In our case, we are only creating Clubhouse stories for `dependabot` PRs. I could see this action being handy for open source maintainers that might be tracking their work through Clubhouse internally, but don't want a new PR from every contributor to flood their Clubhouse backlog.

### CODE REVIEW

**Comment by singingwolfboy** [comment]
Thanks for doing this! I definitely see the use-case, and I agree that this could be a valuable feature. However, this pull request as it's currently written will break backwards compatibility by renaming `ignored-users` to `excluded-users`, and I don't think there's a reason to do that.

What if we call this configuration variable `only-users` instead? That makes it clearer that, if this variable is set, this action will _only_ process pull requests from the provided list of users, instead of from everyone. We can leave the `ignored-users` variable as it is, since it doesn't conflict with the concept of `only-users`.

I also suggest that we refactor the logic here. Rather than having a function called `getIgnoredUsers()`, it now makes more sense to have a function called something like `shouldProcessPullRequestForUser(user)`, which would return true or false. That function would encapsulate all the logic of both `ignored-users` and `only-users`. Essentially, the logic could look roughly like this:

```
if the user is in both "ignored" and "only":
    throw an error
if there is one or more "only" user:
    return whether the user is in "only"
return whether the user is NOT in "ignored"
```

Then you can write automated tests for that function. Be sure to cover _all_ the possible paths: it's important to test for positive cases _and_ negative cases. You'll need to set the `process.env` values for each test, rather than in the `beforeEach`/`afterEach` functions.

Finally, it looks like the built `dist/index.js` file contains different versions of some of the dependencies of this project, which is causing some unrelated changes to appear in the diff. Please make sure that you run `npm install` to get the exact versions of the dependencies, as specified in `package-lock.json`. When you review the diff on GitHub, the only changes visible should be ones related to this `only-users` change. (If you want to change the dependencies of this project, including upgrading them to more recent versions, that's fine but it belongs in a _separate_ pull request, not in this one.)

Does that make sense? Do you have any questions?

**Comment by jdcargile** [comment]
@singingwolfboy, I have made the requested updates.
- Using `only-users` instead of `included-users` to be in better alignment with the backwards compatibility of `ignored-users`
- I've encapsulated all the user list determination in a single `util` method called `shouldProcessPullRequestForUser(user)`
- I have captured all the happy/unhappy paths for `shouldProcessPullRequestForUser(user)`
- I made sure that there were no version bumps in the index file. Thank you for the tip on rerunning `npm install`

File: src/util.ts [hunk 1]
```diff
@@ -237,7 +291,7 @@ export async function createClubhouseStory(
   }
 
   const body: ClubhouseCreateStoryBody = {
-    name: payload.pull_request.title,
+    name: `${payload.repository.name} - ${payload.pull_request.title}`,
```
Comment (singingwolfboy) [inline src/util.ts:294]: This change seems unrelated to the rest. Could you please put it in a separate pull request, so we can discuss it separately?

File: src/util.ts [hunk 2]
```diff
@@ -29,13 +29,67 @@ function stringFromMap(map: Map<Stringable, Stringable>): string {
...
+    );
+    return true;
+  }
+
+  if (onlyUsers && ignoredUsers) {
+    core.setFailed(
+      "You have defined both ignored-users and only-users lists, please use one or the other. Cancelling Clubhouse workflow..."
+    );
+    return false;
+  }
```
Comment (singingwolfboy) [inline src/util.ts:57]: This is only a problem if the same user is in `only-users` and `ignored-users`. I don't see anything wrong with having both inputs specified, as long as they contain different contents.

Also, this should raise an exception, not `return false`.

File: src/util.ts [hunk 3]
```diff
@@ -29,13 +29,67 @@ function stringFromMap(map: Map<Stringable, Stringable>): string {
...
-  const s = new Set<string>();
+export function shouldProcessPullRequestForUser(user: string): boolean {
   const IGNORED_USERS = core.getInput("ignored-users");
-  if (!IGNORED_USERS) {
-    return s;
+  const ONLY_USERS = core.getInput("only-users");
+  let ignoredUsers;
+  let onlyUsers;
+
+  if (IGNORED_USERS) {
```
Comment (singingwolfboy) [inline src/util.ts:38]: There's no need to check and see if these values are truthy or not. If you pass it to the `getUserListAsSet()` function, you'll always get the value you want: a set of the users specified. If we have an empty string, you'll get an empty set -- which is fine! It maintains the variable type, which is nice for static typing.

File: src/util.ts [hunk 4]
```diff
@@ -29,13 +29,67 @@ function stringFromMap(map: Map<Stringable, Stringable>): string {
...
+      return false;
+    } else {
+      core.debug(
+        "PR author is NOT defined in ignored-users list. Proceeding with Clubhouse workflow..."
+      );
+      return true;
+    }
+  }
+
+  return false;
```
Comment (singingwolfboy) [inline src/util.ts:87]: This should return true. If no inputs are specified, this integration should process all pull requests.

File: __tests__/util.test.ts [hunk 5]
```diff
@@ -162,3 +164,33 @@ test("getClubhouseURLFromPullRequest comment", async () => {
...
 });
+
+test("shouldProcessPullRequestForUser user lists not defined", async () => {
+  expect(util.shouldProcessPullRequestForUser("github-author-1")).toBeTruthy;
+});
+
+test("shouldProcessPullRequestForUser both user lists defined", async () => {
+  process.env["INPUT_ONLY-USERS"] = "fake-user-1";
+  process.env["INPUT_IGNORED-USERS"] = "fake-user-1";
+  expect(util.shouldProcessPullRequestForUser("fake-user-1")).toBeFalsy;
```
Comment (singingwolfboy) [inline __tests__/util.test.ts:175]: This should throw an exception.

File: __tests__/util.test.ts [hunk 6]
```diff
@@ -162,3 +164,33 @@ test("getClubhouseURLFromPullRequest comment", async () => {
...
+});
+
+test("shouldProcessPullRequestForUser fake-user-1 in only-users", async () => {
+  process.env["INPUT_ONLY-USERS"] = "fake-user-1";
```
Comment (singingwolfboy) [inline __tests__/util.test.ts:179]: Please make at least one of these tests specify _multiple_ users, to be sure that comma-splitting works correctly. You may want to use [`test.each()`](https://jestjs.io/docs/en/api#testeachtablename-fn-timeout).

**Review by singingwolfboy — CHANGES_REQUESTED** [review]

File: src/util.ts (hunk 2 above)
Comment (jdcargile) [reply src/util.ts:57]: @singingwolfboy, I can account for both lists being populated, but was thinking that is someone is using the `only-users` input, they are explicitly telling the action they only want it to run for the specific users and to ignore all other users. Can you think of a use case where both would be populated?

Comment (singingwolfboy) [reply src/util.ts:57]: It's a bit of an edge case, but I could see someone wanting to try out the new `only-users` input without deleting their existing `ignore-users` input. It's a bit strange to specify both, but as long as they don't actually conflict with each other, I don't see a reason to disallow it.

File: __tests__/util.test.ts (hunk 5 above)
Comment (jdcargile) [reply __tests__/util.test.ts:175]: If I simply remove the `return false`, but keep the `core.setFailed` that should meet the criteria correct? Referencing [GH Action documentation](https://docs.github.com/en/actions/creating-actions/setting-exit-codes-for-actions#setting-a-failure-exit-code-in-a-javascript-action) as I wanted to see how to correctly throw an error.

File: src/util.ts (hunk 1 above)
Comment (jdcargile) [reply src/util.ts:294]: No problem. I'll create a new PR for this.

File: src/util.ts (hunk 4 above)
Comment (jdcargile) [reply src/util.ts:87]: Agreed!

File: __tests__/util.test.ts (hunk 5 above)
Comment (singingwolfboy) [reply __tests__/util.test.ts:175]: `core.setFailed` will set metadata that GitHub Actions uses to determine if the run was successful or not. However, it does not cause the run to halt. If you don't throw an error, it's possible that something in this code will continue to process this pull request in a way we don't want.

Also, conceptually it doesn't make sense to return true _or_ false if we have conflicting information, so we should throw an error instead.

File: src/util.ts (hunk 1 above)
Comment (jdcargile) [reply src/util.ts:294]: No problem. I will create a new PR.

File: __tests__/util.test.ts [hunk 7]
```diff
@@ -162,3 +164,33 @@ test("getClubhouseURLFromPullRequest comment", async () => {
...
+  expect(util.shouldProcessPullRequestForUser("fake-user-1")).toBeTruthy;
+});
+
+test("shouldProcessPullRequestForUser fake-user-1 in ignored-users", async () => {
+  process.env["INPUT_IGNORED-USERS"] = "fake-user-1";
+  expect(util.shouldProcessPullRequestForUser("fake-user-1")).toBeFalsy;
```
Comment (singingwolfboy) [inline __tests__/util.test.ts:185]: All of these `.toBeFalsy` and `.toBeTruthy` calls need parentheses, otherwise they won't be actually _called_. For example:
```suggestion
  expect(util.shouldProcessPullRequestForUser("fake-user-1")).toBeFalsy();
```
See [the Jest docs on matchers](https://jestjs.io/docs/en/using-matchers).

File: src/util.ts (hunk 2 above)
Comment (jdcargile) [reply src/util.ts]: Now throwing an error for the the use case where the author is in both `only-users` and `ignored-users`. Otherwise, I am showing a debug message and proceeding through the logic as requested.

File: __tests__/util.test.ts (hunk 5 above)
Comment (jdcargile) [reply __tests__/util.test.ts]: Fixed.

File: __tests__/util.test.ts (hunk 7 above)
Comment (jdcargile) [reply __tests__/util.test.ts]: Yea, my bad. Thank you for the help and documentation.

**Review by jdcargile — COMMENTED** [review]
@singingwolfboy, I have made the requested updates. I appreciate your patience with my changes and look forward to your review.

**Review by singingwolfboy — APPROVED** [review]
Looks great! Thank you 😄
